### ISSUE TITLE
Domains: a NOT NULL clause in a domain definition ignored

### ISSUE THREAD

**Issue by GoogleCodeExporter** [sagarswathi/h2database#266]
```
Hi!

H2: 1.2.147

This script:
------------------
CREATE DOMAIN DM_ID AS BIGINT NOT NULL;

CREATE TABLE TEST
(
 ID DM_ID
);

ALTER TABLE TEST ADD CONSTRAINT PK_KEY PRIMARY KEY(ID);
------------------

failed with error:
ALTER TABLE test ADD CONSTRAINT PK_KEY PRIMARY KEY(ID);
Column "ID" must not be nullable; SQL statement:
ALTER TABLE TEST ADD CONSTRAINT PK_KEY PRIMARY KEY(ID) [90023-147] 90023/90023.

it seems that primary key does not check domain NOT NULL flag.
```

Original issue reported on code.google.com by `kua...@gmail.com` on 12 Dec 2010 at 1:09

**Comment by GoogleCodeExporter** [comment]
```
I look at INFORMATION_SCHEMA.COLUMNS and it seems that NOT NULL from DOMAIN is 
not propagated to column and it may be a source of problem.
```

Original comment by `kua...@gmail.com` on 12 Dec 2010 at 1:23

**Comment by GoogleCodeExporter** [comment]
```
just one more test case with domain and not null:
-----
create domain dm_test as integer not null;
create table test
( t dm_test);
insert into test (t) values(null);
-----

null is inserted, but should not.
```

Original comment by `kua...@gmail.com` on 13 Dec 2010 at 5:34

**Comment by GoogleCodeExporter** [comment]
```
I dig a little this bug.
This problem in Parser.java  parseColumnForTable() method.
initially column with domain NOT NULL is parsed correctly.
But after that NULL is overrided by defaultNullable parameter.

may be this could be fixed in two ways:
1) add reference to userDataType for column and check 
userDataType.getColumn().isNullable(). As for me this is more right way and it 
may be useful in some other situations concerning domains manipulations.

2) add check for column().isNullable() before overriding by defaultNullable
I mark by >>>>    <<<< markers suggested second patch (one line only).

I hope this helps you to fix this bug. 
----------------------
    private Column parseColumnForTable(String columnName, boolean defaultNullable) {
        Column column;
        boolean isIdentity = false;
        if (readIf("IDENTITY") || readIf("SERIAL")) {
            column = new Column(columnName, Value.LONG);
            column.setOriginalSQL("IDENTITY");
            parseAutoIncrement(column);
            column.setPrimaryKey(true);
        } else {
            column = parseColumn(columnName);
        }
        if (readIf("NOT")) {
            read("NULL");
            column.setNullable(false);
        } else if (readIf("NULL")) {
            column.setNullable(true);
        } else {
          >>>> if (column.isNullable()) <<<<
           column.setNullable(defaultNullable);
        }
.... the rest of code....


```

Original comment by `kua...@gmail.com` on 13 Dec 2010 at 3:09

**Comment by GoogleCodeExporter** [comment]
```
I think the problem is that NOT NULL is always ignored.
Fixed in revision 3234
```

Original comment by `thomas.t...@gmail.com` on 14 Dec 2010 at 8:29
- Changed title: **Domains: a NOT NULL clause in a domain definition ignored**
- Changed state: **Accepted**

**Comment by GoogleCodeExporter** [comment]
```
Fixed in version 1.3.149
```

Original comment by `thomas.t...@gmail.com` on 10 Jan 2011 at 7:21
- Changed state: **Fixed**
